commit 6cde16f3f4711583ae4d896dfcc125d25c7d5b0d
Author:     Zhihong Zhang <100308595+nvidianz@users.noreply.github.com>
AuthorDate: Mon Aug 8 09:37:33 2022 -0400
Commit:     GitHub <noreply@github.com>
CommitDate: Mon Aug 8 09:37:33 2022 -0400

    1. Replaced Pickle with FOBS (#746)
    
    2. Added decomposers for all classes to be serialized
    3. Changed all examples to use FOBS
    4. Break all recursion in the serialized objects.
---
 nvflare/private/fed/client/scheduler_cmds.py | 12 ++++++------
 nvflare/private/fed/server/server_engine.py  | 21 ++++++++++++---------
 2 files changed, 18 insertions(+), 15 deletions(-)

diff --git a/nvflare/private/fed/client/scheduler_cmds.py b/nvflare/private/fed/client/scheduler_cmds.py
index 4ed2668b..099abbb9 100644
--- a/nvflare/private/fed/client/scheduler_cmds.py
+++ b/nvflare/private/fed/client/scheduler_cmds.py
@@ -12,12 +12,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import pickle
 from typing import List
 
 from nvflare.apis.fl_constant import ReturnCode, SystemComponents
 from nvflare.apis.resource_manager_spec import ResourceConsumerSpec, ResourceManagerSpec
 from nvflare.apis.shareable import Shareable
+from nvflare.fuel.utils import fobs
 from nvflare.private.admin_defs import Message
 from nvflare.private.defs import ERROR_MSG_PREFIX, RequestHeader, TrainingTopic
 from nvflare.private.fed.client.admin import RequestProcessor
@@ -42,7 +42,7 @@ class CheckResourceProcessor(RequestProcessor):
         with engine.new_context() as fl_ctx:
             result = Shareable()
             try:
-                resource_spec = pickle.loads(req.body)
+                resource_spec = fobs.loads(req.body)
                 check_result, token = resource_manager.check_resources(
                     resource_requirement=resource_spec, fl_ctx=fl_ctx
                 )
@@ -51,7 +51,7 @@ class CheckResourceProcessor(RequestProcessor):
             except Exception:
                 result.set_return_code(ReturnCode.EXECUTION_EXCEPTION)
 
-        return Message(topic="reply_" + req.topic, body=pickle.dumps(result))
+        return Message(topic="reply_" + req.topic, body=fobs.dumps(result))
 
 
 class StartJobProcessor(RequestProcessor):
@@ -76,7 +76,7 @@ class StartJobProcessor(RequestProcessor):
 
         try:
             with engine.new_context() as fl_ctx:
-                resource_spec = pickle.loads(req.body)
+                resource_spec = fobs.loads(req.body)
                 job_id = req.get_header(RequestHeader.JOB_ID)
                 token = req.get_header(ShareableHeader.RESOURCE_RESERVE_TOKEN)
                 allocated_resources = resource_manager.allocate_resources(
@@ -115,10 +115,10 @@ class CancelResourceProcessor(RequestProcessor):
             result = Shareable()
             try:
                 # resource_spec = req.get_header(ShareableHeader.RESOURCE_SPEC)
-                resource_spec = pickle.loads(req.body)
+                resource_spec = fobs.loads(req.body)
                 token = req.get_header(ShareableHeader.RESOURCE_RESERVE_TOKEN)
                 resource_manager.cancel_resources(resource_requirement=resource_spec, token=token, fl_ctx=fl_ctx)
             except Exception:
                 result.set_return_code(ReturnCode.EXECUTION_EXCEPTION)
 
-        return Message(topic="reply_" + req.topic, body=pickle.dumps(result))
+        return Message(topic="reply_" + req.topic, body=fobs.dumps(result))
diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index ca229c8f..20917fbb 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -16,7 +16,6 @@ import copy
 import logging
 import multiprocessing
 import os
-import pickle
 import re
 import shlex
 import shutil
@@ -52,6 +51,7 @@ from nvflare.apis.utils.common_utils import get_open_ports
 from nvflare.apis.utils.fl_context_utils import get_serializable_data
 from nvflare.apis.workspace import Workspace
 from nvflare.fuel.hci.zip_utils import zip_directory_to_bytes
+from nvflare.fuel.utils import fobs
 from nvflare.private.admin_defs import Message
 from nvflare.private.defs import RequestHeader, TrainingTopic
 from nvflare.private.fed.server.server_json_config import ServerJsonConfigurator
@@ -72,8 +72,11 @@ class ClientConnection:
         self.client = client
 
     def send(self, data):
-        data = pickle.dumps(data)
-        self.client.send(data)
+        try:
+            buf = fobs.dumps(data)
+        except BaseException as e:
+            raise e
+        self.client.send(buf)
 
     def recv(self):
         return self.client.recv()
@@ -573,7 +576,7 @@ class ServerEngine(ServerEngineInternalSpec):
         # Send the aux messages through admin_server
         request.set_peer_props(fl_ctx.get_all_public_props())
 
-        message = Message(topic=ReservedTopic.AUX_COMMAND, body=pickle.dumps(request))
+        message = Message(topic=ReservedTopic.AUX_COMMAND, body=fobs.dumps(request))
         message.set_header(RequestHeader.JOB_ID, str(fl_ctx.get_prop(FLContextKey.CURRENT_RUN)))
         requests = {}
         for n in targets:
@@ -585,7 +588,7 @@ class ServerEngine(ServerEngineInternalSpec):
             client_name = self.get_client_name_from_token(r.client_token)
             if r.reply:
                 try:
-                    results[client_name] = pickle.loads(r.reply.body)
+                    results[client_name] = fobs.loads(r.reply.body)
                 except BaseException:
                     results[client_name] = make_reply(ReturnCode.COMMUNICATION_ERROR)
                     self.logger.error(
@@ -708,7 +711,7 @@ class ServerEngine(ServerEngineInternalSpec):
             # assume server resource is unlimited
             if site_name == "server":
                 continue
-            request = Message(topic=TrainingTopic.CHECK_RESOURCE, body=pickle.dumps(resource_requirements))
+            request = Message(topic=TrainingTopic.CHECK_RESOURCE, body=fobs.dumps(resource_requirements))
             client = self.get_client_from_name(site_name)
             if client:
                 requests.update({client.token: request})
@@ -719,7 +722,7 @@ class ServerEngine(ServerEngineInternalSpec):
         for r in replies:
             site_name = self.get_client_name_from_token(r.client_token)
             if r.reply:
-                resp = pickle.loads(r.reply.body)
+                resp = fobs.loads(r.reply.body)
                 result[site_name] = (
                     resp.get_header(ShareableHeader.CHECK_RESOURCE_RESULT, False),
                     resp.get_header(ShareableHeader.RESOURCE_RESERVE_TOKEN, ""),
@@ -736,7 +739,7 @@ class ServerEngine(ServerEngineInternalSpec):
             check_result, token = result
             if check_result and token:
                 resource_requirements = resource_reqs[site_name]
-                request = Message(topic=TrainingTopic.CANCEL_RESOURCE, body=pickle.dumps(resource_requirements))
+                request = Message(topic=TrainingTopic.CANCEL_RESOURCE, body=fobs.dumps(resource_requirements))
                 request.set_header(ShareableHeader.RESOURCE_RESERVE_TOKEN, token)
                 client = self.get_client_from_name(site_name)
                 if client:
@@ -749,7 +752,7 @@ class ServerEngine(ServerEngineInternalSpec):
         for site, dispatch_info in client_sites.items():
             resource_requirement = dispatch_info.resource_requirements
             token = dispatch_info.token
-            request = Message(topic=TrainingTopic.START_JOB, body=pickle.dumps(resource_requirement))
+            request = Message(topic=TrainingTopic.START_JOB, body=fobs.dumps(resource_requirement))
             request.set_header(RequestHeader.JOB_ID, job_id)
             request.set_header(ShareableHeader.RESOURCE_RESERVE_TOKEN, token)
             client = self.get_client_from_name(site)
